// starknetkit/dist/index-68877418.js
var p = Object.defineProperty;
var f = (e, n, t) => n in e ? p(e, n, { enumerable: !0, configurable: !0, writable: !0, value: t }) : e[n] = t;
var l = (e, n, t) => (f(e, typeof n != "symbol" ? n + "" : n, t), t);
import { constants as I } from "starknet";
import { C as m, a as o, b as i, d as v } from "./connector-ca2f4153.js";
const r = "f2e613881f7a0e811295cdd57999e31b", w = `<svg
    width="32"
    height="32"
    viewBox="0 0 32 32"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect width="32" height="32" rx="8" fill="#FF875B" />
    <path
      d="M18.316 8H13.684C13.5292 8 13.4052 8.1272 13.4018 8.28531C13.3082 12.7296 11.0323 16.9477 7.11513 19.9355C6.99077 20.0303 6.96243 20.2085 7.05335 20.3369L9.76349 24.1654C9.85569 24.2957 10.0353 24.3251 10.1618 24.2294C12.6111 22.3734 14.5812 20.1345 16 17.6529C17.4187 20.1345 19.389 22.3734 21.8383 24.2294C21.9646 24.3251 22.1443 24.2957 22.2366 24.1654L24.9467 20.3369C25.0375 20.2085 25.0092 20.0303 24.885 19.9355C20.9676 16.9477 18.6918 12.7296 18.5983 8.28531C18.5949 8.1272 18.4708 8 18.316 8Z"
      fill="white"
    />
  </svg>`, y = () => Object.keys(localStorage).some(
  (e) => e === "walletconnect" || e.startsWith("wc@2:")
), h = () => {
  if (y()) {
    delete localStorage.walletconnect;
    for (const e in localStorage)
      e.startsWith("wc@2:") && delete localStorage[e];
  }
}, A = (e) => {
  localStorage.setItem("starknetLastConnectedWallet", e);
}, b = () => {
  localStorage.removeItem("starknetLastConnectedWallet");
}, L = () => window?.starknet_argentX ? window?.starknet_argentX?.isInAppBrowser : !1;
class N extends m {
  constructor(t = {}) {
    super();
    l(this, "_wallet", null);
    l(this, "_options");
    this._options = t;
  }
  available() {
    return !0;
  }
  async ready() {
    return this._wallet ? this._wallet.isPreauthorized() : !1;
  }
  get id() {
    return "argentMobile";
  }
  get name() {
    return "Argent (mobile)";
  }
  get icon() {
    return {
      dark: w,
      light: w
    };
  }
  get wallet() {
    if (!this._wallet)
      throw new o();
    return this._wallet;
  }
  async connect() {
    if (await this.ensureWallet(), !this._wallet)
      throw new i();
    const t = this._wallet.account, a = await this.chainId();
    return {
      account: t.address,
      chainId: a
    };
  }
  async disconnect() {
    if (await this._wallet.disable(), h(), !this.available() && !this._wallet)
      throw new i();
    if (!this._wallet?.isConnected)
      throw new v();
    this._wallet = null;
  }
  async account() {
    if (!this._wallet || !this._wallet.account)
      throw new o();
    return this._wallet.account;
  }
  async chainId() {
    if (!this._wallet || !this.wallet.account || !this._wallet.provider)
      throw new o();
    const t = await this._wallet.provider.getChainId();
    return BigInt(t);
  }
  // needed, methods required by starknet-react. Otherwise an exception is throwd
  async initEventListener(t) {
    if (!this._wallet)
      throw new o();
    this._wallet.on("accountsChanged", t);
  }
  // needed, methods required by starknet-react. Otherwise an exception is throwd
  async removeEventListener(t) {
    if (!this._wallet)
      throw new o();
    this._wallet.off("accountsChanged", t), this._wallet = null;
  }
  async ensureWallet() {
    const { getStarknetWindowObject: t } = await import("./index-c78ebd52.js"), { chainId: a, projectId: s, dappName: d, description: u, url: _ } = this._options, g = {
      chainId: a ?? I.NetworkName.SN_MAIN,
      name: d,
      projectId: s ?? r,
      description: u,
      url: _
    };
    s === r && (console.log("========= NOTICE ========="), console.log(
      "While your application will continue to function, we highly recommended"
    ), console.log("signing up for your own API keys."), console.log(
      "Go to WalletConnect Cloud (https://cloud.walletconnect.com) and create a new account."
    ), console.log(
      "Once your account is created, create a new project and collect the Project ID"
    ), console.log("=========================="));
    const C = await t(g);
    this._wallet = C;
    const c = this._wallet;
    await c.enable(), c.client.on("session_delete", () => {
      h(), this._wallet = null, b(), document.dispatchEvent(new Event("wallet_disconnected"));
    });
  }
}
export {
  N as A,
  h as a,
  L as i,
  b as r,
  A as s
};
