// @mantine/notifications/cjs/Notifications.cjs
'use client';
'use strict';

var React = require('react');
var reactTransitionGroup = require('react-transition-group');
var core = require('@mantine/core');
var hooks = require('@mantine/hooks');
var getNotificationStateStyles = require('./get-notification-state-styles.cjs');
var NotificationContainer = require('./NotificationContainer.cjs');
var notifications_store = require('./notifications.store.cjs');
var Notifications_module = require('./Notifications.module.css.cjs');

const Transition = reactTransitionGroup.Transition;
const defaultProps = {
  position: "bottom-right",
  autoClose: 4e3,
  transitionDuration: 250,
  containerWidth: 440,
  notificationMaxHeight: 200,
  limit: 5,
  zIndex: core.getDefaultZIndex("overlay"),
  store: notifications_store.notificationsStore,
  withinPortal: true
};
const varsResolver = core.createVarsResolver(
  (_, { zIndex, position, containerWidth }) => {
    const [vertical, horizontal] = position.split("-");
    return {
      root: {
        "--notifications-z-index": zIndex?.toString(),
        "--notifications-top": vertical === "top" ? "var(--mantine-spacing-md)" : void 0,
        "--notifications-bottom": vertical === "bottom" ? "var(--mantine-spacing-md)" : void 0,
        "--notifications-left": horizontal === "left" ? "var(--mantine-spacing-md)" : horizontal === "center" ? "50%" : void 0,
        "--notifications-right": horizontal === "right" ? "var(--mantine-spacing-md)" : void 0,
        "--notifications-transform": horizontal === "center" ? "translateX(-50%)" : void 0,
        "--notifications-container-width": core.rem(containerWidth)
      }
    };
  }
);
const Notifications = core.factory((_props, ref) => {
  const props = core.useProps("Notifications", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    position,
    autoClose,
    transitionDuration,
    containerWidth,
    notificationMaxHeight,
    limit,
    zIndex,
    store,
    portalProps,
    withinPortal,
    ...others
  } = props;
  const theme = core.useMantineTheme();
  const data = notifications_store.useNotifications(store);
  const forceUpdate = hooks.useForceUpdate();
  const shouldReduceMotion = hooks.useReducedMotion();
  const refs = React.useRef({});
  const previousLength = React.useRef(0);
  const reduceMotion = theme.respectReducedMotion ? shouldReduceMotion : false;
  const duration = reduceMotion ? 1 : transitionDuration;
  const getStyles = core.useStyles({
    name: "Notifications",
    classes: Notifications_module,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  React.useEffect(() => {
    store?.updateState((current) => ({ ...current, limit: limit || 5 }));
  }, [limit]);
  hooks.useDidUpdate(() => {
    if (data.notifications.length > previousLength.current) {
      setTimeout(() => forceUpdate(), 0);
    }
    previousLength.current = data.notifications.length;
  }, [data.notifications]);
  const items = data.notifications.map(({ style: notificationStyle, ...notification }) => /* @__PURE__ */ React.createElement(
    Transition,
    {
      key: notification.id,
      timeout: duration,
      onEnter: () => refs.current[notification.id].offsetHeight,
      nodeRef: { current: refs.current[notification.id] }
    },
    (state) => /* @__PURE__ */ React.createElement(
      NotificationContainer.NotificationContainer,
      {
        ref: (node) => {
          refs.current[notification.id] = node;
        },
        data: notification,
        onHide: (id) => notifications_store.hideNotification(id, store),
        autoClose,
        ...getStyles("notification", {
          style: {
            ...getNotificationStateStyles.getNotificationStateStyles({
              state,
              position,
              transitionDuration: duration,
              maxHeight: notificationMaxHeight
            }),
            ...notificationStyle
          }
        })
      }
    )
  ));
  return /* @__PURE__ */ React.createElement(core.OptionalPortal, { withinPortal, ...portalProps }, /* @__PURE__ */ React.createElement(core.Box, { ...getStyles("root"), ref, ...others }, /* @__PURE__ */ React.createElement(reactTransitionGroup.TransitionGroup, null, items)));
});
Notifications.classes = Notifications_module;
Notifications.displayName = "@mantine/notifications/Notifications";
Notifications.show = notifications_store.notifications.show;
Notifications.hide = notifications_store.notifications.hide;
Notifications.update = notifications_store.notifications.update;
Notifications.clean = notifications_store.notifications.clean;
Notifications.cleanQueue = notifications_store.notifications.cleanQueue;
Notifications.updateState = notifications_store.notifications.updateState;

exports.Notifications = Notifications;
