// @mantine/dates/esm/components/PickerInputBase/PickerInputBase.mjs
'use client';
import React from 'react';
import cx from 'clsx';
import { factory, useInputProps, CloseButton, Modal, Input, Popover } from '@mantine/core';
import { HiddenDatesInput } from '../HiddenDatesInput/HiddenDatesInput.mjs';
import classes from './PickerInputBase.module.css.mjs';

const defaultProps = {};
const PickerInputBase = factory((_props, ref) => {
  const {
    inputProps,
    wrapperProps,
    placeholder,
    classNames,
    styles,
    unstyled,
    popoverProps,
    modalProps,
    dropdownType,
    children,
    formattedValue,
    dropdownHandlers,
    dropdownOpened,
    onClick,
    clearable,
    onClear,
    clearButtonProps,
    rightSection,
    shouldClear,
    readOnly,
    disabled,
    value,
    name,
    form,
    type,
    ...others
  } = useInputProps("PickerInputBase", defaultProps, _props);
  const _rightSection = rightSection || (clearable && shouldClear && !readOnly && !disabled ? /* @__PURE__ */ React.createElement(
    CloseButton,
    {
      variant: "transparent",
      onClick: onClear,
      unstyled,
      size: inputProps.size || "sm",
      ...clearButtonProps
    }
  ) : null);
  const handleClose = () => {
    const isInvalidRangeValue = type === "range" && Array.isArray(value) && value[0] && !value[1];
    if (isInvalidRangeValue) {
      onClear();
    }
    dropdownHandlers.close();
  };
  return /* @__PURE__ */ React.createElement(React.Fragment, null, dropdownType === "modal" && !readOnly && /* @__PURE__ */ React.createElement(
    Modal,
    {
      opened: dropdownOpened,
      onClose: handleClose,
      withCloseButton: false,
      size: "auto",
      "data-dates-modal": true,
      unstyled,
      ...modalProps
    },
    children
  ), /* @__PURE__ */ React.createElement(Input.Wrapper, { ...wrapperProps }, /* @__PURE__ */ React.createElement(
    Popover,
    {
      position: "bottom-start",
      opened: dropdownOpened,
      onClose: handleClose,
      disabled: dropdownType === "modal" || readOnly,
      trapFocus: true,
      returnFocus: true,
      unstyled,
      ...popoverProps
    },
    /* @__PURE__ */ React.createElement(Popover.Target, null, /* @__PURE__ */ React.createElement(
      Input,
      {
        "aria-label": formattedValue || placeholder,
        "data-dates-input": true,
        "data-read-only": readOnly || void 0,
        disabled,
        component: "button",
        type: "button",
        multiline: true,
        onClick: (event) => {
          onClick?.(event);
          dropdownHandlers.toggle();
        },
        rightSection: _rightSection,
        ...inputProps,
        ref,
        classNames: { ...classNames, input: cx(classes.input, classNames?.input) },
        ...others
      },
      formattedValue || /* @__PURE__ */ React.createElement(Input.Placeholder, { error: inputProps.error, unstyled }, placeholder)
    )),
    /* @__PURE__ */ React.createElement(Popover.Dropdown, { "data-dates-dropdown": true }, children)
  )), /* @__PURE__ */ React.createElement(HiddenDatesInput, { value, name, form, type }));
});
PickerInputBase.classes = classes;
PickerInputBase.displayName = "@mantine/dates/PickerInputBase";

export { PickerInputBase };
