// trpc-browser/link/window.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.windowLink = void 0;
const base_1 = require("./internal/base");
const windowLink = (opts) => {
    var _a;
    const handlerMap = new Map();
    const listenWindow = opts.window;
    const postWindow = (_a = opts.postWindow) !== null && _a !== void 0 ? _a : listenWindow;
    return (0, base_1.createBaseLink)({
        postMessage(message) {
            postWindow.postMessage(message, {
                targetOrigin: opts.postOrigin,
            });
        },
        addMessageListener(listener) {
            const handler = (ev) => {
                listener(ev.data);
            };
            handlerMap.set(listener, handler);
            listenWindow.addEventListener('message', handler);
        },
        removeMessageListener(listener) {
            const handler = handlerMap.get(listener);
            if (handler) {
                listenWindow.removeEventListener('message', handler);
            }
        },
        addCloseListener(listener) {
            listenWindow.addEventListener('beforeunload', listener);
        },
        removeCloseListener(listener) {
            listenWindow.removeEventListener('beforeunload', listener);
        },
    });
};
exports.windowLink = windowLink;
